### PR TITLE
Add Lua API for Prometheus Metrics

### PR DESCRIPTION
**Pull request by rubenwardy**
A while back, nerzhul [added support for prometheus](https://github.com/minetest/minetest/pull/9719).

This PR adds an API to allow mods to add their own Prometheus metrics. They can create multi-dimensional counter or gauge metrics.

Here's what I'm using it for: https://monitor.rubenwardy.com/d/9TgIegyGk/ctf

## To do

This PR is a Work in Progress.

- [x] Support for labels
- [ ] Clean up Metric type hierarchy/API
- [ ] Clean up how dimensional data is created (no point in passing in type and help again)
- [ ] Testing

## How to test

```lua
if not minetest.create_metric then
	error("No metrics!")
	return
end

local chat_counter = minetest.create_metric("counter", "chat_messages", "Total chat messages")
minetest.register_on_chat_message(function()
	chat_counter:increment()
end)
```

### CODE REVIEW

File: src/map.h [hunk 1]
```diff
@@ -428,7 +428,7 @@ class ServerMap : public Map
 	MapDatabase *dbase = nullptr;
 	MapDatabase *dbase_ro = nullptr;
 
-	MetricCounterPtr m_save_time_counter;
```
Comment (nerzhul) [inline src/map.h:431]: warning, you have both MetricCounter & MetricGauche in prometheus. If you remove the meaning here, we don't know what is this

File: src/util/metricsbackend.h [hunk 2]
```diff
@@ -122,17 +122,19 @@ class SimpleMetricGauge : public MetricGauge
...
 public:
 	MetricsBackend() = default;
 
 	virtual ~MetricsBackend() {}
 
-	virtual MetricCounterPtr addCounter(
-			const std::string &name, const std::string &help_str);
-	virtual MetricGaugePtr addGauge(
-			const std::string &name, const std::string &help_str);
+	virtual MetricPtr addCounter(const std::string &name, const std::string &help_str,
```
Comment (nerzhul) [inline src/util/metricsbackend.h:134]: if you remove the Counter, it's implicit, and it's not implicit in any other prometheus API. Counter is not generic

**Review by nerzhul — CHANGES_REQUESTED** [review]

**Comment by rubenwardy** [comment]
Should I add an empty class MetricCounter for this, which derives from Metric? Seems unnecessary. I need a unified interface in order to cleanly add MetricRef. Counter has all the base behaviour, it's gauge which then allows set() and decrement()

**Comment by nerzhul** [comment]
There is no histogram or summary in prometheus, it's done by the queries on the prometheus server, it's why it's simple.
Ok for the rename.
Can you look at adding binding to remove metrics ? In some cases it's useful. For example in a such case:

metric_a{label_a="toto", state="running"}=1
metric_a{label_a="toto", state="stopped"}=1
You must drop one of the metric in order to setup the other metric labeled to 1, or set to zero but it's not in the best practices

**Review by nerzhul — CHANGES_REQUESTED** [review]
see my comment before

**Comment by erlehmann** [comment]
@rubenwardy why was this closed?

**Comment by rubenwardy** [comment]
> @rubenwardy why was this closed?

A lot of effort for something that won't effect many users and doesn't seem that supported

**Pull request closed without merging**